src/consensus/aft/raft.h

```diff
@@ -1259,51 +1342,253 @@ namespace aft
+      AsyncExecutionRet(
+        Aft<LedgerProxy, ChannelProxy, SnapshotterProxy>* self_) :
+        self(self_)
+      {}
+
+      Aft<LedgerProxy, ChannelProxy, SnapshotterProxy>* self;
+    };
+
     static void execute_append_entries_cb(
       std::unique_ptr<threading::Tmsg<AsyncExecution>> msg)
     {
-      msg->data.self->execute_append_entries(
-        msg->data.append_entries, msg->data.r, msg->data.confirm_evidence);
-      msg->cb =
-        reinterpret_cast<void (*)(std::unique_ptr<threading::ThreadMsg>)>(
-          continue_execution);
+      auto self = msg->data.self;
+      std::unique_lock<SpinLock> guard(self->state->lock);
+      self->execute_append_entries(std::move(msg));
+    }
+
+    void execute_append_entries(
+      std::unique_ptr<threading::Tmsg<AsyncExecution>> msg)
+    {
+      auto self = msg->data.self;
+      if (self->consensus_type == ConsensusType::CFT)
+      {
+        self->execute_append_entries_sync(msg);
+      }
+      else
+      {
+        bool result = self->execute_append_entries_async(msg);
+        if (!result)
+        {
+          return;
+        }
+      }
+
+      auto msg_ret = std::make_unique<threading::Tmsg<AsyncExecutionRet>>(
+        continue_execution, self);
       if (threading::ThreadMessaging::thread_count > 1)
       {
         threading::ThreadMessaging::thread_messaging.add_task(
-          threading::MAIN_THREAD_ID, std::move(msg));
+          threading::MAIN_THREAD_ID, std::move(msg_ret));
       }
       else
       {
-        msg->cb(std::move(msg));
+        msg_ret->cb(std::move(msg_ret));
       }
     }
 
     static void continue_execution(
-      std::unique_ptr<threading::Tmsg<AsyncExecution>> msg)
+      std::unique_ptr<threading::Tmsg<AsyncExecutionRet>> msg)
     {
       msg->data.self->is_execution_pending = false;
       msg->data.self->try_execute_pending();
     }
 
-    void execute_append_entries(
+    struct AsyncExecutionCtx
+    {
+      AsyncExecutionCtx() : pending_cbs(0) {}
+
+      uint32_t pending_cbs;
+    };
+
+    struct AsyncExecTxMsg
+    {
+      AsyncExecTxMsg(
+        Aft<LedgerProxy, ChannelProxy, SnapshotterProxy>* self_,
+        std::unique_ptr<kv::AbstractExecutionWrapper>&& ds_,
+        std::unique_ptr<threading::Tmsg<AsyncExecution>>&& msg_,
+        kv::Version last_idx_,
+        kv::Version commit_idx_,
+        uint16_t home_thread_) :
+        self(self_),
+        ds(std::move(ds_)),
+        msg(std::move(msg_)),
+        last_idx(last_idx_),
+        commit_idx(commit_idx_),
+        home_thread(home_thread_)
+      {}
+
+      Aft<LedgerProxy, ChannelProxy, SnapshotterProxy>* self;
+      std::unique_ptr<kv::AbstractExecutionWrapper> ds;
+      std::unique_ptr<threading::Tmsg<AsyncExecution>> msg;
+      kv::Version last_idx;
+      kv::Version commit_idx;
+      uint16_t home_thread;
+      std::shared_ptr<AsyncExecutionCtx> ctx;
+    };
+
+    uint64_t next_exec_thread = 0;
+
+    bool execute_append_entries_sync(
```

Comment:
That's a 200+ lines function, it's quite unfortunate, is there a way we can break this down?